firedrake/preconditioners/gtmg.py

```diff
@@ -0,0 +1,197 @@
+        if not get_coarse_operator:
+            raise ValueError("Need to provide a callback which provides the coarse operator.")
+        coarse_operator = get_coarse_operator()
+
+        coarse_space_bcs = appctx.get("coarse_space_bcs", None)
+
+        # These should be callbacks which return the relevant nullspaces
+        get_coarse_nullspace = appctx.get("get_coarse_op_nullspace", None)
+        get_coarse_transpose_nullspace = appctx.get("get_coarse_op_transpose_nullspace", None)
+
+        self.coarse_op = allocate_matrix(coarse_operator,
+                                         bcs=coarse_space_bcs,
+                                         form_compiler_parameters=fcp,
+                                         mat_type=coarse_mat_type,
+                                         options_prefix=coarse_options_prefix)
+        self._assemble_coarse_op = create_assembly_callable(coarse_operator,
+                                                            tensor=self.coarse_op,
+                                                            bcs=coarse_space_bcs,
+                                                            form_compiler_parameters=fcp)
+        self._assemble_coarse_op()
+        self.coarse_op.force_evaluation()
+        coarse_opmat = self.coarse_op.petscmat
+
+        # Set nullspace if provided
+        if get_coarse_nullspace:
+            nsp = get_coarse_nullspace()
+            coarse_opmat.setNullSpace(nsp.nullspace(comm=pc.comm))
+
+        if get_coarse_transpose_nullspace:
+            tnsp = get_coarse_transpose_nullspace()
+            coarse_opmat.setTransposeNullSpace(tnsp.nullspace(comm=pc.comm))
+
+        interp_petscmat = appctx.get("interpolation_matrix", None)
+        if interp_petscmat is None:
+            # Create interpolation matrix from coarse space to fine space
+            fine_space = ctx.J.arguments()[0].function_space()
+            interpolator = Interpolator(TestFunction(coarse_space), fine_space)
+            interpolation_matrix = interpolator.callable()
+            interpolation_matrix._force_evaluation()
+            interp_petscmat = interpolation_matrix.handle
+
+        # We set up a PCMG object that uses the constructed interpolation
+        # matrix to generate the restriction/prolongation operators.
+        # This is a two-level multigrid preconditioner.
+        pcmg = PETSc.PC().create(comm=pc.comm)
+        pcmg.incrementTabLevel(1, parent=pc)
+
+        pcmg.setType(pc.Type.MG)
+        pcmg.setOptionsPrefix(options_prefix)
+        pcmg.setMGLevels(2)
+        pcmg.setMGCycleType(pc.MGCycleType.V)
+        pcmg.setMGInterpolation(1, interp_petscmat)
+        pcmg.setOperators(A=fine_petscmat, P=fine_petscmat)
+
+        # We set a DM and an appropriate SNESContext for the coarse solver
+        # so we can do multigrid or patch solves.
+        from firedrake.variational_solver import NonlinearVariationalProblem
```

Comment:
Lines 143 -- 156 are definitely a hack. I think @wence- had a suggestion for improving this, but I have forgotten what was said. I think a utility method for `PCBase` which generates an appropriate SNES context for all subclassed PC objects would remove this hack (and prevent this snippet from being duplicated in `firedrake.SCPC` and `firedrake.HybridizationPC`).